exporter/metrics_test.go

```diff
@@ -2,14 +2,98 @@ package exporter
 
 import (
 	"testing"
+	"time"
 
-	"gotest.tools/assert"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/stretchr/testify/assert"
+	"go.mongodb.org/mongo-driver/bson/primitive"
 )
 
-func TestMakeRawMetricName(t *testing.T) {
-	for name, fqName := range map[string]string{
-		"dd.serverStatus.opcountersRepl.delete": "dd_server_status_opcounters_repl_delete",
-	} {
-		assert.Equal(t, fqName, makeRawMetricName(name))
+// Test metric renaming and labeling.
+func TestMetricRename(t *testing.T) {
+	tcs := []struct {
+		prefix     string
+		name       string
+		wantMetric string
+		wantLabel  string
+	}{
+		{
+			prefix:     "serverStatus.metrics.commands.saslStart.",
+			name:       "total",
+			wantMetric: "mongodb_ss_metrics_commands_saslStart_total",
+		},
+		{
+			prefix:     "serverStatus.metrics.commands._configsvrShardCollection.",
+			name:       "failed",
+			wantMetric: "mongodb_ss_metrics_commands_configsvrShardCollection_failed",
+		},
+		{
+			prefix:     "serverStatus.wiredTiger.lock.",
+			name:       "metadata lock acquisitions",
+			wantMetric: "mongodb_ss_wt_lock_metadata_lock_acquisitions",
+		},
+		{
+			prefix:     "serverStatus.wiredTiger.perf.",
+			name:       "file system write latency histogram (bucket 5) - 500-999ms",
+			wantMetric: "mongodb_ss_wt_perf",
+			wantLabel:  "perf_bucket",
+		},
+		{
+			prefix:     "serverStatus.wiredTiger.transaction.",
+			name:       "rollback to stable updates removed from lookaside",
+			wantMetric: "mongodb_ss_wt_txn_rollback_to_stable_updates_removed_from_lookaside",
+		},
 	}
+
+	for _, tc := range tcs {
+		metric, label := rename(tc.prefix, tc.name)
+		assert.Equal(t, tc.wantMetric, metric, tc.prefix+tc.name)
+		assert.Equal(t, tc.wantLabel, label, tc.prefix+tc.name)
+	}
+}
+
+// Test supported value types conversion.
+func TestMakeRawMetric(t *testing.T) {
+	prefix := "serverStatus.transactions."
+	name := "retriedCommandsCount"
+	testCases := []struct {
+		value     interface{}
+		wantVal   *float64
+		wantError bool
+	}{
+		{value: true, wantVal: pointerTo(1), wantError: false},
+		{value: false, wantVal: pointerTo(0), wantError: false},
+		{value: int32(1), wantVal: pointerTo(1), wantError: false},
+		{value: int64(2), wantVal: pointerTo(2), wantError: false},
+		{value: float32(1.23), wantVal: pointerTo(float64(float32(1.23))), wantError: false},
+		{value: float64(1.23), wantVal: pointerTo(1.23), wantError: false},
+		{value: time.Date(2020, 06, 15, 0, 0, 0, 0, time.UTC), wantVal: nil, wantError: true},
+		{value: primitive.A{}, wantVal: nil, wantError: false},
+		{value: primitive.Timestamp{}, wantVal: nil, wantError: false},
+		{value: "zapp", wantVal: nil, wantError: false},
+		{value: []byte{}, wantVal: nil, wantError: true},
+	}
+
+	fqName := prometheusize(prefix + name)
+	ln := make([]string, 0)
```

Comment:
`var ln []string` ?

Reply:
if we are going to use go-consistent:
```
go-consistent ./...
./exporter/metrics.go:148:8: empty slice: use make([]T, 0)
./exporter/metrics.go:149:8: empty slice: use make([]T, 0)
```

Reply:
`go-consistent` calculates a number of different patterns and insists on the most common/consistent one. But we really want to use `var x []T`, and if we change that _everywhere_, it will not complain.